cpp/src/jni/dataset/jni_util.cc

```diff
@@ -167,9 +167,16 @@ ReservationListenableMemoryPool::~ReservationListenableMemoryPool() {}
 
 Status CheckException(JNIEnv* env) {
   if (env->ExceptionCheck()) {
-    env->ExceptionDescribe();
+    jthrowable t = env->ExceptionOccurred();
     env->ExceptionClear();
-    return Status::Invalid("Error during calling Java code from native code");
+    jclass describer_class =
+        env->FindClass("org/apache/arrow/dataset/jni/JniExceptionDescriber");
+    jmethodID describe_method = env->GetStaticMethodID(
+        describer_class, "describe", "(Ljava/lang/Throwable;)Ljava/lang/String;");
+    std::string description = JStringToCString(
+        env, (jstring)env->CallStaticObjectMethod(describer_class, describe_method, t));
+    return Status::Invalid("Error during calling Java code from native code: " +
+                           description);
```

Comment:
Hmm... I don't know what is possible exactly, but ideally we should:
1) map Java exception classes to specific C++ status codes (for example a `IOException` would map to `Status::IOError` while `InvalidTypeException` would map to `Status::TypeError`)
2) keep the Java exception object in a Java-specific `StatusDetail` subclass, just like we do [for Python exceptions](https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/common.cc#L54-L139)

cc @lwhite1

Reply:
Also this helper function should be moved out of dataset-specific code to be reused generically in all JNI bindings.

Reply:
> Also this helper function should be moved out of dataset-specific code to be reused generically in all JNI bindings.

Was considering doing the similar thing but maybe after we have a common code module for JNI stuffs?

Reply:
@pitrou sure, I have opened a ticket
https://issues.apache.org/jira/browse/ARROW-16691

Reply:
@zhztheplayer Do you want to address the other parts of the initial comment in this PR?

Reply:
Addressed. Thanks for suggestion.